### PR TITLE
Acyl Feature: Refactor UI API Docs

### PR DESCRIPTION
**Pull request by esterlinej**
## Changes
- replace bootprint with base openapi-generator
- refactor yaml from swagger to OpenAPI 3.0.3 spec
- update path from /swagger to /apidocs
- add/update schemas, responses, securitySchemes
- replace broken inheritance with static schemas
- uses generated html2 output

### Test Local
Build Acyl locally, run Acyl MockUI, and navigate to the API Docs from the /help page:
1. `$ make`
2. `$ acyl mockui --ui-enforce-oauth --ui-path ./ui`
3. Browse to /help and select API Documentation

### CODE REVIEW

**Comment by esterlinej** [comment]
Screen Shot 2020-12-16 at 7 12 35 PM
Screen Shot 2020-12-16 at 7 13 14 PM

**Review by bkeroackdsc — CHANGES_REQUESTED** [review]
- You probably should add explicit API key permissions to each endpoint (IIRC, only the DELETE env endpoint requires admin key permissions).
- `/apidocs` is better than `/doc`

**Comment by esterlinej** [comment]
Update to output html2, provides updated view
Screen Shot 2020-12-17 at 4 14 44 PM
Screen Shot 2020-12-17 at 4 14 55 PM

File: doc/generate/openapi.yml [hunk 1]
```diff
@@ -28,77 +35,391 @@ tags:
...
+    github_auth:
+      description: github repository permission level access
+      type: oauth2
+      flows:
+        implicit:
+          authorizationUrl: "https://github.com/oauth/login"
+          scopes:
+            read: requires read api key privileges
+            write: requires write api key privileges
+            admin: requires admin api key privileges
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:60]: Omit. This shouldn't be used by API consumers. In the future, UI json API endpoints may also support API key auth.

File: doc/generate/openapi.yml [hunk 2]
```diff
@@ -28,77 +35,391 @@ tags:
...
+
+security:
+  - api_key: []
+  - github_auth:
+      - read
+
+components:
+  securitySchemes:
+    api_key:
+      description: API Key with required github_auth permission level or aborts with Unauthorized
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:47]: This isn't "github_auth" permission level, it's the permission level of the api key when it was created.

File: doc/generate/openapi.yml [hunk 3]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      flows:
+        implicit:
+          authorizationUrl: "https://github.com/oauth/login"
+          scopes:
+            read: requires read api key privileges
+            write: requires write api key privileges
+            admin: requires admin api key privileges
+  schemas:
+    QAEnvironmentEvent:
+      description: A representation of QAEnvironmentEvent object
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:63]: This description is tautological ("X is a representation of X").

Give a useful description: "This is data representing an event associated with an environment object"

File: doc/generate/openapi.yml [hunk 4]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      type: object
+      properties:
+        timestamp:
+          type: string
+          format: date-time
+        message:
+          type: string
+    RefMap:
+      description: A representation of RefMap object; map of repo to git reference
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:72]: "Map of repo name to git reference (branch)."

File: doc/generate/openapi.yml [hunk 5]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      deprecated: true
+      description: "**DEPRECATED** not supported with Nitro v2 or newer; A representation of AminoServiceToPort object"
+      type: object
+      properties:
+        string:
+          type: integer
+          format: int64
+    QAEnvironment_v0:
+      type: object
+      description: A representation of a v0 environment
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:87]: What is a "v0 environment"? Be more descriptive.

File: doc/generate/openapi.yml [hunk 6]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      properties:
+        name:
+          type: string
+        created:
+          type: string
+          format: date-time
+        raw_events:
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:94]: I'm not sure if this actually is part of the JSON payload. Double check.

File: doc/generate/openapi.yml [hunk 7]
```diff
@@ -28,77 +35,391 @@ tags:
...
+          type: array
+          items:
+            type: string
+          format: string
+        events:
+          $ref: '#/components/schemas/QAEnvironmentEvent'
+        hostname:
+          type: string
+        qa_type:
+          type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:104]: I'm pretty sure these are deprecated & unused.

File: doc/generate/openapi.yml [hunk 8]
```diff
@@ -28,77 +35,391 @@ tags:
...
+        pull_request:
+          type: integer
+        source_sha:
+          type: string
+        base_sha:
+          type: string
+        source_branch:
+          type: string
+        base_branch:
+          type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:118]: Call out that these are fields of the triggering Pull Request. Explain what a "base SHA/branch" and "source SHA/branch" are (and why they are different).

File: doc/generate/openapi.yml [hunk 9]
```diff
@@ -28,77 +35,391 @@ tags:
...
+        raw_status:
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:119]: I don't believe this is part of the JSON response.

File: doc/generate/openapi.yml [hunk 10]
```diff
@@ -28,77 +35,391 @@ tags:
...
+          type: string
+        status:
+          type: integer
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:122]: Explain that this is an integer-based enum and give the possible values (X: success, Y: failed, etc).

File: doc/generate/openapi.yml [hunk 11]
```diff
@@ -28,77 +35,391 @@ tags:
...
+        ref_map:
+          $ref: '#/components/schemas/RefMap'
+        amino_service_to_port:
+          type: object
+        amino_kubernetes_namespace:
+          type: string
+        amino_environment_id:
+          type: integer
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:130]: Deprecated, unused. API clients should ignore.

File: doc/generate/openapi.yml [hunk 12]
```diff
@@ -28,77 +35,391 @@ tags:
...
+    QAEnvironment_v1:
+      type: object
+      description: A representation of a v1 environment
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:133]: Summarize the differences between a v0 environment response and a v1 response. See the API changelog for details.

File: doc/generate/openapi.yml [hunk 13]
```diff
@@ -28,77 +35,391 @@ tags:
...
+        ref_map:
+          $ref: '#/components/schemas/RefMap'
+        amino_service_to_port:
+          type: object
+        amino_kubernetes_namespace:
+          type: string
+        amino_environment_id:
+          type: integer
+        commit_sha_map:
+          $ref: '#/components/schemas/RefMap'
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:178]: Commit SHA map is different from a ref map. Refmap uses branch names, commit sha map uses full commit shas.

File: doc/generate/openapi.yml [hunk 14]
```diff
@@ -28,77 +35,391 @@ tags:
...
+    QAEnvironment_v2:
+      type: object
+      description: A representation of a v2 environment
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:181]: Summarizes changes from v1.

File: doc/generate/openapi.yml [hunk 15]
```diff
@@ -28,77 +35,391 @@ tags:
...
+        amino_environment_id:
+          deprecated: true
+          type: integer
+          description: "**DEPRECATED** not supported with Nitro v2 or newer"
+        commit_sha_map:
+          $ref: '#/components/schemas/RefMap'
+        source_ref:
+          type: string
+    EventLog:
+      description: A representation of an event log response
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:235]: Describe more fully what this is, and what an event log contains. Why is it useful?

File: doc/generate/openapi.yml [hunk 16]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      type: object
+      properties:
+        id:
+          type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:239]: UUID.

File: doc/generate/openapi.yml [hunk 17]
```diff
@@ -28,77 +35,391 @@ tags:
...
+        repo:
+          type: string
+        pull_request:
+          type: integer
+        webhook_payload:
+          type: string
+          format: byte
+        log:
+          type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:256]: Explain that this is a time-ordered array of debug log messages associated with this specific event.

File: doc/generate/openapi.yml [hunk 18]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      required: true
+      schema:
+        type: string
+    daysParam:
+      name: days
+      in: query
+      description: number of days (default `1`)
+      required: false
+      schema:
+        type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:320]: integer

File: doc/generate/openapi.yml [hunk 19]
```diff
@@ -28,77 +35,391 @@ tags:
...
+    eventLogIdParam:
+      name: id
+      in: path
+      description: event log id
+      required: true
+      schema:
+        type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:327]: UUID

File: doc/generate/openapi.yml [hunk 20]
```diff
@@ -28,77 +35,391 @@ tags:
...
+      required: false
+      schema:
+        type: string
+    statusParam:
+      name: status
+      in: query
+      description: enviornment status
+      required: false
+      schema:
+        type: string
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:285]: integer (IIRC)

**Review by bkeroackdsc — CHANGES_REQUESTED** [review]

File: doc/generate/openapi.yml (hunk 7 above)
Comment (esterlinej) [reply doc/generate/openapi.yml:92]: They still exist in the response struct, but I can add `deprecated: true` to this field

File: doc/generate/openapi.yml (hunk 10 above)
Comment (esterlinej) [reply doc/generate/openapi.yml:99]: Screen Shot 2020-12-17 at 7 47 29 PM

**Comment by esterlinej** [comment]
Updated most descriptions
Screen Shot 2020-12-17 at 8 48 54 PM

**Comment by esterlinej** [comment]
Made one final push updating summary and description text, also adding missing update to status in the screen cap above.

**Comment by esterlinej** [comment]
Screen Shot 2020-12-17 at 7 47 29 PM

File: doc/generate/openapi.yml [hunk 21]
```diff
@@ -0,0 +1,979 @@
...
+        pull_request:
+          type: integer
+          format: uint
+          description: "GitHub Pull Request ID"
+        source_sha:
+          type: string
+          description: "GitHub SHA for the dev branch"
+        base_sha:
+          type: string
+          description: "GitHub SHA for the master branch"
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:103]: "GitHub HEAD commit SHA for the branch the PR will merge into (eg, master)."

File: doc/generate/openapi.yml [hunk 22]
```diff
@@ -0,0 +1,979 @@
...
+        source_branch:
+          type: string
+          description: "GitHub name of the dev branch"
+        base_branch:
+          type: string
+          description: "GitHub name of the master branch"
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:109]: "GitHub branch name the PR will merge into (eg, master)."

File: doc/generate/openapi.yml [hunk 23]
```diff
@@ -0,0 +1,979 @@
...
+      description: "This data represents the default environment object response"
+      properties:
+        timestamp:
+          type: string
+          format: date-time
+        message:
+          type: string
+    RefMap:
+      type: object
+      description: "This data represents a map of repo name to git reference (branch) object response"
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:61]: Remove the word "data" and the phrase "object response" from each of these descriptions.

File: doc/generate/openapi.yml [hunk 24]
```diff
@@ -0,0 +1,979 @@
...
+          $ref: '#/components/schemas/QAEnvironmentEvent'
+        hostname:
+          type: string
+        qa_type:
+          deprecated: true
+          type: string
+          description: "**DEPRECATED** not supported with Nitro v2 or newer"
+        user:
+          type: string
+          description: "GitHub username of the requesting user"
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:90]: "GitHub username of the user who opened the triggering PR."

File: doc/generate/openapi.yml [hunk 25]
```diff
@@ -0,0 +1,979 @@
...
+                items:
+                  type: object
+                  $ref: '#/components/schemas/QAEnvironment_v0'
+        500:
+          description: "Internal Server Error"
+  /envs/_search:
+    get:
+      tags:
+        - backward_compatible
+      summary: "Search environments from optional parameters, at least one parameter is required"
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:420]: Point out that multiple parameters are combined with a logical AND. PR = X *and* repo = Y *and* status = Z, etc.

File: doc/generate/openapi.yml [hunk 26]
```diff
@@ -0,0 +1,979 @@
...
+          schema:
+            type: string
+      responses:
+        204:
+          description: "No Content"
+        404:
+          description: "Not Found"
+        500:
+          description: "Internal Server Error"
+  /envs/{name}/success:
```
Comment (bkeroackdsc) [inline doc/generate/openapi.yml:519]: I second thought, I think you can get rid of all these old endpoints. They almost certainly don't do anything and nobody should be trying to use them. They're just clutter at this point.

**Review by bkeroackdsc — CHANGES_REQUESTED** [review]

File: doc/generate/openapi.yml (hunk 26 above)
Comment (bkeroackdsc) [reply doc/generate/openapi.yml:519]: I mean, remove them from the API docs. Later on we can remove them from the codebase as well.

**Review by bkeroackdsc — APPROVED** [review]

**Pull request merged**